Source/DafnyCore/Resolver/PreType/PreTypeStateWithErrorMessage.cs

```diff
@@ -0,0 +1,37 @@
+//-----------------------------------------------------------------------------
+//
+// Copyright by the contributors to the Dafny Project
+// SPDX-License-Identifier: MIT
+//
+//-----------------------------------------------------------------------------
+
+using System;
+using System.Diagnostics.Contracts;
+
+namespace Microsoft.Dafny {
+  abstract class PreTypeStateWithErrorMessage {
```

Comment:
I didn't understand your reply to my previous comment, but the name of this class doesn't make sense to me, both on its own and in the way it is used. Constraints and state seem like two different things to me.